doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,139 @@
+- RDM: rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md
+- Title: Runtime configuration architecture
+- Authors: José Álamos, Leandro Lanzieri
+- Status: draft
+- Type: Design
+- Created: December 2018
+
+## Abstract
+This memo describes the proposed high-level architecture and
+mechanisms to implement a runtime configuration system on a RIOT node.
+
+A runtime configuration system is in charge of providing a mechanism to set and
+get the values of configuration parameters that are used during the execution of the firmware,
+as well as a way to persist these values. Most of times these values are
+per-node specific.
+
+Examples of runtime configurations are:
+- Transmission duty cycles
+- Sensor thresholds
+- Private-public keys
+- System state variables
+
+These parameters might have constraints, like an specific order to be applied
+(due to interdependencies) or value boundaries.
+
+The main advantages of having such a system are:
+- Easy to apply per-node configuration during deployment
+- No need to implement a special mechanism for per-node configurations during
+  firmware updates (only in the case of migration)
+- Common interface for modules to expose their runtime configuration parameters and handle
+  them
+- Transparent interface for storing configuration parameters in non-volatile
+  storage devices
+
+## Status
+This document is currently under open discussion. This document is a product of
+[Configuration Task
+Force](https://github.com/RIOT-OS/RIOT/wiki/Configuration-Task-Force-(CTF)), and
+aims to describe the architecture of a runtime configuration system. The content
+of this document is licensed with a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) terminology
+and the following acronyms:
+
+- RDM: RIOT Developer Memo
+- CTF: RIOT's Configuration Task Force
+- RCS: Runtime Configuration System
+
+# 1. Introduction
+This document specifies the proposed architecture by the
+Configuration Task Force (CFT) to implement a secure and reliable Runtime
+Configuration System (RCS), focusing in modularity, reusing of existing
+technologies in RIOT (network stack, storage interface) and adhering to
+standards.
+
+# 2. Architecture
+The RCS is formed by [the RIOT Registry](3-the-riot-registry), one or more [Registry Handlers](4-registry-handlers), one or more
+[storage facilities](5-storage-facilities) and a [configuration manager](6-configuration-manager) with one or more interfaces. The
```

Comment:
Could you add a bit more explanation here? Commonly, an architecture consists of components, their _semantics_, their _interfaces_, and their _interactions_. I would love to see an exhaustive high-level description here. In particular, I haven't really caught:

- Is the registry just a facade or does it provide intrinsic functionality? 
- Is the Registry an order-preserving key-value store of contextually grouped keys? 
- What is the API of the Registry in relation to the APIs of the Registry Handlers?
- Is the configuration manager actively handling system configurations? Or is it just a skeleton for orchestrating the access to the configuration system?
- What does a  Registry Handler do beyond providing a contextual name?
- How do all these components interact when retrieving or changing a configuration parameter?

Reply:
> Could you add a bit more explanation here? Commonly, an architecture consists of components, their semantics, their interfaces, and their interactions. I would love to see an exhaustive high-level description here. In particular, I haven't really caught:

Sure, we will.

I will try to answer here anyway and rephrase this section on the fly.

> Is the registry just a facade or does it provide intrinsic functionality?

The registry provides intrinsic functionality:
- Register and keep a list of _registry handlers_
- Orchestrate config operations via these _registry handlers_ (get, set, commit, export)

> Is the Registry an order-preserving key-value store of contextually grouped keys?

I would say the Registry is an _orchestrator_ to configure RIOT modules at runtime (via _registry handlers_) and persist these configurations in any data structures like file systems (via _storage_).

The combination of the Registry + Registry Handlers + Storage Facilities provide a key-value store of contextually grouped keys (ordered or non-ordered) at code level. Then, a Configuration Manager is required e.g to configure a TX interval via CoAP.

> What is the API of the Registry in relation to the APIs of the Registry Handlers?

The API of the Registry provides high level functions to handle persistent configurations. The API of the Registry Handlers provides low level functions that are intended to be orchestrated by the Registry.

In RIOT slang, the Registry is to `netapi` as the Registry Handlers are to `netdev`.

> Is the configuration manager actively handling system configurations? Or is it just a skeleton for orchestrating the access to the configuration system?

As described before, it's closer to a skeleton for orchestrating the access to the configuration system.

> What does a Registry Handler do beyond providing a contextual name?

A Registry Handler also provides 4 functions required to do all configuration operations, as described in [4. Registry Handlers](https://github.com/RIOT-OS/RIOT/pull/10622/files#diff-63356eb67626b5872bf0edb05a1a0befR77). [Here's a conceptual example of how a registry handler would look](https://gist.github.com/jia200x/cca922092ca0d13957dee203e9ad9b55)

> How do all these components interact when retrieving or changing a configuration parameter?

E.g for retrieving the value of "a_config" (uint16_t) from a module "module" in a "buf" variable, a user would do:
```
registry_get_value("module/a_config", buf, sizeof(uint16_t));
```
This:
1. Searches for a Registry Handler with "module" name in the global list of registry handlers.
2. If found, calls the registry handler's `get` function with "a_config" key.
3. The `get` function copies the value to the `buf` variable

E.g for changing a configuration parameter:
```
/* Set "a_config" to the value in buf */
registry_set_value("module/a_config", buf);
```
This:
1. Searches for a Registry Handler with "module" name in the global list of registry handlers.
2. If found, calls the registry handler's `set` function with "a_config" key.
3. The `set` function copies the value in the `buf` variable to the corresponding variable.

At any time it's possible to either load or save all configuration parameters from the non-volatile storage device with dedicated load and save functions.
E.g:
```
/* At this point all configuration have their default values */

registry_load();

/* Now, all configurations got their values from the non-volatile storage */

registry_set_value("module/a_config", buf, sizeof(uint16_t));

/* At this point "a_config" has a different value BUT STILL NOT REFLECTED IN THE STORAGE

registry_save();

/* Now, the new configuration is stored permanently */
```

As one could imagine, `registry_load` will read all configurations from the storage and call all registry handler `set` functions . And of course, `registry_save` will call all registry handler `get` functions and save them in the storage.

Hope everything is more clear now. I will reflect these answers in the RDM. Please don't hesitate to ask again if needed.